File: Kernel Simulator/Misc/Screensaver/Displays/Glitch.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Namespace Misc.Screensaver.Displays
    Public Module GlitchSettings

        Private _GlitchDelay As Integer = 10
        Private _GlitchDensity As Integer = 40

        ''' <summary>
        ''' [Glitch] How many milliseconds to wait before making the next write?
        ''' </summary>
        Public Property GlitchDelay As Integer
            Get
                Return _GlitchDelay
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 10
                _GlitchDelay = value
            End Set
        End Property
        ''' <summary>
        ''' [Glitch] The Glitch density in percent
        ''' </summary>
        Public Property GlitchDensity As Integer
            Get
                Return _GlitchDensity
            End Get
            Set(value As Integer)
                If value < 0 Then value = 40
                If value > 100 Then value = 40
                _GlitchDensity = value
            End Set
        End Property

    End Module

    Public Class GlitchDisplay
        Inherits BaseScreensaver
        Implements IScreensaver

        Private RandomDriver As Random
        Private CurrentWindowWidth As Integer
        Private CurrentWindowHeight As Integer
        Private ResizeSyncing As Boolean

        Public Overrides Property ScreensaverName As String = "Glitch" Implements IScreensaver.ScreensaverName

        Public Overrides Property ScreensaverSettings As Dictionary(Of String, Object) Implements IScreensaver.ScreensaverSettings

        Public Overrides Sub ScreensaverPreparation() Implements IScreensaver.ScreensaverPreparation
            'Variable preparations
            RandomDriver = New Random
            CurrentWindowWidth = ConsoleWrapper.WindowWidth
            CurrentWindowHeight = ConsoleWrapper.WindowHeight
            Console.BackgroundColor = ConsoleColor.Black
            Console.ForegroundColor = ConsoleColor.White
            ConsoleWrapper.CursorVisible = False
            ConsoleWrapper.Clear()
        End Sub

        Public Overrides Sub ScreensaverLogic() Implements IScreensaver.ScreensaverLogic
            'Select random positions to generate the glitch
            Dim GlitchDense As Double = If(GlitchDensity > 100, 100, GlitchDensity) / 100
            Dim AmountOfBlocks As Integer = ConsoleWrapper.WindowWidth * ConsoleWrapper.WindowHeight
            Dim BlocksToCover As Integer = AmountOfBlocks * GlitchDense
            Dim CoveredBlocks As New ArrayList
            Do Until CoveredBlocks.Count = BlocksToCover Or ResizeSyncing
                If CurrentWindowHeight <> ConsoleWrapper.WindowHeight Or CurrentWindowWidth <> ConsoleWrapper.WindowWidth Then ResizeSyncing = True
                If Not ResizeSyncing Then
                    Dim CoverX As Integer = RandomDriver.Next(ConsoleWrapper.WindowWidth)
                    Dim CoverY As Integer = RandomDriver.Next(ConsoleWrapper.WindowHeight)
                    ConsoleWrapper.SetCursorPosition(CoverX, CoverY)

                    'Select random glitch type
                    Dim GlitchType As GlitchType = [Enum].Parse(GetType(GlitchType), RandomDriver.Next(5))

                    'Select random letter
                    Dim LetterCapitalized As Boolean = RandomDriver.Next(2)
                    Dim LetterRangeStart As Integer = If(LetterCapitalized, 65, 97)
                    Dim LetterRangeEnd As Integer = If(LetterCapitalized, 90, 122)
                    Dim Letter As Char = Convert.ToChar(RandomDriver.Next(LetterRangeStart, LetterRangeEnd + 1))

                    'Select random symbol
                    Dim UseExtendedAscii As Boolean = RandomDriver.Next(2)
                    Dim SymbolRangeStart As Integer = If(UseExtendedAscii, 128, 33)
                    Dim SymbolRangeEnd As Integer = If(UseExtendedAscii, 256, 64)
                    Dim Symbol As Char = Convert.ToChar(RandomDriver.Next(SymbolRangeStart, SymbolRangeEnd + 1))

                    'Select red, green, or blue background and foreground
                    Dim GlitchBlockColorType As GlitchColorType = [Enum].Parse(GetType(GlitchColorType), RandomDriver.Next(3))
                    Dim GlitchLetterColorType As GlitchColorType = [Enum].Parse(GetType(GlitchColorType), RandomDriver.Next(3))
                    Dim ColorLetter As Boolean = RandomDriver.Next(2)
                    Dim ColorBlockNumber As Integer = RandomDriver.Next(0, 256)
                    Dim ColorLetterNumber As Integer = RandomDriver.Next(0, 256)
                    Dim ColorBlockInstance As Color = Color.Empty
                    Dim ColorLetterInstance As Color = Color.Empty

                    '...    for the block
                    Select Case GlitchBlockColorType
                        Case GlitchColorType.Red
                            ColorBlockInstance = New Color(ColorBlockNumber, 0, 0)
                        Case GlitchColorType.Green
                            ColorBlockInstance = New Color(0, ColorBlockNumber, 0)
                        Case GlitchColorType.Blue
                            ColorBlockInstance = New Color(0, 0, ColorBlockNumber)
                    End Select

                    '...and for the letter
                    Select Case GlitchLetterColorType
                        Case GlitchColorType.Red
                            ColorLetterInstance = New Color(ColorLetterNumber, 0, 0)
                        Case GlitchColorType.Green
                            ColorLetterInstance = New Color(0, ColorLetterNumber, 0)
                        Case GlitchColorType.Blue
                            ColorLetterInstance = New Color(0, 0, ColorLetterNumber)
                    End Select

                    'Now, print based on the glitch type
                    Select Case GlitchType
                        Case GlitchType.RandomLetter
                            If ColorLetter Then SetConsoleColor(ColorLetterInstance) Else Console.ForegroundColor = ConsoleColor.White
                            WritePlain(Letter, False)
                        Case GlitchType.RandomSymbol
                            If ColorLetter Then SetConsoleColor(ColorLetterInstance) Else Console.ForegroundColor = ConsoleColor.White
                            WritePlain(Symbol, False)
                        Case GlitchType.RedGreenBlueColor
                            SetConsoleColor(ColorBlockInstance, True)
                            WritePlain(" ", False)
                        Case GlitchType.RedGreenBlueColorWithRandomLetter
                            If ColorLetter Then SetConsoleColor(ColorLetterInstance) Else Console.ForegroundColor = ConsoleColor.White
                            SetConsoleColor(ColorBlockInstance, True)
                            WritePlain(Letter, False)
                        Case GlitchType.RedGreenBlueColorWithRandomSymbol
                            If ColorLetter Then SetConsoleColor(ColorLetterInstance) Else Console.ForegroundColor = ConsoleColor.White
                            SetConsoleColor(ColorBlockInstance, True)
                            WritePlain(Symbol, False)
                    End Select
                    If Not CoveredBlocks.Contains(CStr(CoverX) + ", " + CStr(CoverY)) Then CoveredBlocks.Add(CStr(CoverX) + ", " + CStr(CoverY))
                Else
                    'We're resizing.
                    ConsoleWrapper.CursorVisible = False
                    Exit Do
                End If
                SleepNoBlock(GlitchDelay, ScreensaverDisplayerThread)
            Loop

            'Reset resize sync
            ResizeSyncing = False
            CurrentWindowWidth = ConsoleWrapper.WindowWidth
            CurrentWindowHeight = ConsoleWrapper.WindowHeight
        End Sub

    End Class

    Enum GlitchType
        ''' <summary>
        ''' A block with either red, green, or blue as color, and can be darkened
        ''' </summary>
        RedGreenBlueColor
        ''' <summary>
        ''' A block with either red, green, or blue as color, and can be darkened and feature a random letter printed in white
        ''' </summary>
        RedGreenBlueColorWithRandomLetter
        ''' <summary>
        ''' A block with either red, green, or blue as color, and can be darkened and feature a random symbol printed in white
        ''' </summary>
        RedGreenBlueColorWithRandomSymbol
        ''' <summary>
        ''' A random letter printed in white, red, green, or blue, and can be darkened
        ''' </summary>
        RandomLetter
        ''' <summary>
        ''' A random symbol printed in white, red, green, or blue, and can be darkened
        ''' </summary>
        RandomSymbol
    End Enum

    Enum GlitchColorType
        ''' <summary>
        ''' The red color
        ''' </summary>
        Red
        ''' <summary>
        ''' The green color
        ''' </summary>
        Green
        ''' <summary>
        ''' The blue color
        ''' </summary>
        Blue
    End Enum
End Namespace

File: Kernel Simulator/Misc/Screensaver/Displays/Glitch.cs
using System;
using System.Collections;
using System.Collections.Generic;
using KS.ConsoleBase.Colors;
using KS.Misc.Threading;
using KS.Misc.Writers.ConsoleWriters;
using Terminaux.Base;
using Terminaux.Colors;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

namespace KS.Misc.Screensaver.Displays
{
	public static class GlitchSettings
	{

		private static int _GlitchDelay = 10;
		private static int _GlitchDensity = 40;

		/// <summary>
		/// [Glitch] How many milliseconds to wait before making the next write?
		/// </summary>
		public static int GlitchDelay
		{
			get
			{
				return _GlitchDelay;
			}
			set
			{
				if (value <= 0)
					value = 10;
				_GlitchDelay = value;
			}
		}
		/// <summary>
		/// [Glitch] The Glitch density in percent
		/// </summary>
		public static int GlitchDensity
		{
			get
			{
				return _GlitchDensity;
			}
			set
			{
				if (value < 0)
					value = 40;
				if (value > 100)
					value = 40;
				_GlitchDensity = value;
			}
		}

	}

	public class GlitchDisplay : BaseScreensaver, IScreensaver
	{

		private Random RandomDriver;
		private int CurrentWindowWidth;
		private int CurrentWindowHeight;
		private bool ResizeSyncing;

		public override string ScreensaverName { get; set; } = "Glitch";

		public override Dictionary<string, object> ScreensaverSettings { get; set; }

		public override void ScreensaverPreparation()
		{
			// Variable preparations
			RandomDriver = new Random();
			CurrentWindowWidth = ConsoleWrapper.WindowWidth;
			CurrentWindowHeight = ConsoleWrapper.WindowHeight;
			Console.BackgroundColor = ConsoleColor.Black;
			Console.ForegroundColor = ConsoleColor.White;
			ConsoleWrapper.CursorVisible = false;
			ConsoleWrapper.Clear();
		}

		public override void ScreensaverLogic()
		{
			// Select random positions to generate the glitch
			double GlitchDense = (GlitchSettings.GlitchDensity > 100 ? 100 : GlitchSettings.GlitchDensity) / 100d;
			int AmountOfBlocks = ConsoleWrapper.WindowWidth * ConsoleWrapper.WindowHeight;
			int BlocksToCover = (int)Math.Round(AmountOfBlocks * GlitchDense);
			var CoveredBlocks = new ArrayList();
			while (!(CoveredBlocks.Count == BlocksToCover | ResizeSyncing))
			{
				if (CurrentWindowHeight != ConsoleWrapper.WindowHeight | CurrentWindowWidth != ConsoleWrapper.WindowWidth)
					ResizeSyncing = true;
				if (!ResizeSyncing)
				{
					int CoverX = RandomDriver.Next(ConsoleWrapper.WindowWidth);
					int CoverY = RandomDriver.Next(ConsoleWrapper.WindowHeight);
					ConsoleWrapper.SetCursorPosition(CoverX, CoverY);

					// Select random glitch type
					GlitchType GlitchType = (GlitchType)Convert.ToInt32(Enum.Parse(typeof(GlitchType), RandomDriver.Next(5).ToString()));

					// Select random letter
					bool LetterCapitalized = Convert.ToBoolean(RandomDriver.Next(2));
					int LetterRangeStart = LetterCapitalized ? 65 : 97;
					int LetterRangeEnd = LetterCapitalized ? 90 : 122;
					char Letter = Convert.ToChar(RandomDriver.Next(LetterRangeStart, LetterRangeEnd + 1));

					// Select random symbol
					bool UseExtendedAscii = Convert.ToBoolean(RandomDriver.Next(2));
					int SymbolRangeStart = UseExtendedAscii ? 128 : 33;
					int SymbolRangeEnd = UseExtendedAscii ? 256 : 64;
					char Symbol = Convert.ToChar(RandomDriver.Next(SymbolRangeStart, SymbolRangeEnd + 1));

					// Select red, green, or blue background and foreground
					GlitchColorType GlitchBlockColorType = (GlitchColorType)Convert.ToInt32(Enum.Parse(typeof(GlitchColorType), RandomDriver.Next(3).ToString()));
					GlitchColorType GlitchLetterColorType = (GlitchColorType)Convert.ToInt32(Enum.Parse(typeof(GlitchColorType), RandomDriver.Next(3).ToString()));
					bool ColorLetter = Convert.ToBoolean(RandomDriver.Next(2));
					int ColorBlockNumber = RandomDriver.Next(0, 256);
					int ColorLetterNumber = RandomDriver.Next(0, 256);
					var ColorBlockInstance = Color.Empty;
					var ColorLetterInstance = Color.Empty;

					// ...    for the block
					switch (GlitchBlockColorType)
					{
						case GlitchColorType.Red:
							{
								ColorBlockInstance = new Color(ColorBlockNumber, 0, 0);
								break;
							}
						case GlitchColorType.Green:
							{
								ColorBlockInstance = new Color(0, ColorBlockNumber, 0);
								break;
							}
						case GlitchColorType.Blue:
							{
								ColorBlockInstance = new Color(0, 0, ColorBlockNumber);
								break;
							}
					}

					// ...and for the letter
					switch (GlitchLetterColorType)
					{
						case GlitchColorType.Red:
							{
								ColorLetterInstance = new Color(ColorLetterNumber, 0, 0);
								break;
							}
						case GlitchColorType.Green:
							{
								ColorLetterInstance = new Color(0, ColorLetterNumber, 0);
								break;
							}
						case GlitchColorType.Blue:
							{
								ColorLetterInstance = new Color(0, 0, ColorLetterNumber);
								break;
							}
					}

					// Now, print based on the glitch type
					switch (GlitchType)
					{
						case Displays.GlitchType.RandomLetter:
							{
								if (ColorLetter)
									KernelColorTools.SetConsoleColor(ColorLetterInstance);
								else
									Console.ForegroundColor = ConsoleColor.White;
								TextWriterColor.WritePlain(Convert.ToString(Letter), false);
								break;
							}
						case Displays.GlitchType.RandomSymbol:
							{
								if (ColorLetter)
									KernelColorTools.SetConsoleColor(ColorLetterInstance);
								else
									Console.ForegroundColor = ConsoleColor.White;
								TextWriterColor.WritePlain(Convert.ToString(Symbol), false);
								break;
							}
						case Displays.GlitchType.RedGreenBlueColor:
							{
								KernelColorTools.SetConsoleColor(ColorBlockInstance, true);
								TextWriterColor.WritePlain(" ", false);
								break;
							}
						case Displays.GlitchType.RedGreenBlueColorWithRandomLetter:
							{
								if (ColorLetter)
									KernelColorTools.SetConsoleColor(ColorLetterInstance);
								else
									Console.ForegroundColor = ConsoleColor.White;
								KernelColorTools.SetConsoleColor(ColorBlockInstance, true);
								TextWriterColor.WritePlain(Convert.ToString(Letter), false);
								break;
							}
						case Displays.GlitchType.RedGreenBlueColorWithRandomSymbol:
							{
								if (ColorLetter)
									KernelColorTools.SetConsoleColor(ColorLetterInstance);
								else
									Console.ForegroundColor = ConsoleColor.White;
								KernelColorTools.SetConsoleColor(ColorBlockInstance, true);
								TextWriterColor.WritePlain(Convert.ToString(Symbol), false);
								break;
							}
					}
					if (!CoveredBlocks.Contains(CoverX.ToString() + ", " + CoverY.ToString()))
						CoveredBlocks.Add(CoverX.ToString() + ", " + CoverY.ToString());
				}
				else
				{
					// We're resizing.
					ConsoleWrapper.CursorVisible = false;
					break;
				}
				ThreadManager.SleepNoBlock(GlitchSettings.GlitchDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
			}

			// Reset resize sync
			ResizeSyncing = false;
			CurrentWindowWidth = ConsoleWrapper.WindowWidth;
			CurrentWindowHeight = ConsoleWrapper.WindowHeight;
		}

	}

	enum GlitchType
	{
		/// <summary>
		/// A block with either red, green, or blue as color, and can be darkened
		/// </summary>
		RedGreenBlueColor,
		/// <summary>
		/// A block with either red, green, or blue as color, and can be darkened and feature a random letter printed in white
		/// </summary>
		RedGreenBlueColorWithRandomLetter,
		/// <summary>
		/// A block with either red, green, or blue as color, and can be darkened and feature a random symbol printed in white
		/// </summary>
		RedGreenBlueColorWithRandomSymbol,
		/// <summary>
		/// A random letter printed in white, red, green, or blue, and can be darkened
		/// </summary>
		RandomLetter,
		/// <summary>
		/// A random symbol printed in white, red, green, or blue, and can be darkened
		/// </summary>
		RandomSymbol
	}

	enum GlitchColorType
	{
		/// <summary>
		/// The red color
		/// </summary>
		Red,
		/// <summary>
		/// The green color
		/// </summary>
		Green,
		/// <summary>
		/// The blue color
		/// </summary>
		Blue
	}
}